src/Cli/dotnet/commands/dotnet-test/MSBuildUtility.cs

```diff
@@ -0,0 +1,218 @@
+        {
+            var projectCollection = new ProjectCollection();
+            bool isBuiltOrRestored = true;
+
+            if (!buildOptions.HasNoRestore)
+            {
+                isBuiltOrRestored = BuildOrRestoreProjectOrSolution(
+                    projectFilePath,
+                    projectCollection,
+                    buildOptions,
+                    [CliConstants.RestoreCommand]);
+            }
+
+            if (!buildOptions.HasNoBuild)
+            {
+                isBuiltOrRestored = isBuiltOrRestored && BuildOrRestoreProjectOrSolution(
+                    projectFilePath,
+                    projectCollection,
+                    buildOptions,
+                    [CliConstants.BuildCommand]);
+            }
+
+            var allProjects = GetRelatedProjects(projectFilePath, projectCollection);
+
+            return (allProjects, isBuiltOrRestored);
+        }
+
+        private static bool BuildOrRestoreProjectOrSolution(string filePath, ProjectCollection projectCollection, BuildOptions buildOptions, string[] commands)
+        {
+            var parameters = GetBuildParameters(projectCollection, buildOptions);
+            var globalProperties = GetGlobalProperties(buildOptions);
+
+            var buildRequestData = new BuildRequestData(filePath, globalProperties, null, commands, null);
+
+            BuildResult buildResult = BuildManager.DefaultBuildManager.Build(parameters, buildRequestData);
+
+            return buildResult.OverallResult == BuildResultCode.Success;
+        }
+
+        private static List<Module> GetRelatedProjects(string solutionOrProjectFilePath, ProjectCollection projectCollection)
+        {
+            var allProjects = new List<Module>();
+            IEnumerable<Module> relatedProjects = SolutionAndProjectUtility.GetProjectPropertiesInternal(solutionOrProjectFilePath, projectCollection);
+            allProjects.AddRange(relatedProjects);
+
+            return allProjects;
+        }
+
+        private static ConcurrentBag<Module> ProcessProjectsInParallel(ProjectCollection projectCollection, IEnumerable<string> projects, int degreeOfParallelism)
+        {
+            var allProjects = new ConcurrentBag<Module>();
+
+            Parallel.ForEach(
+                projects,
+                new ParallelOptions { MaxDegreeOfParallelism = degreeOfParallelism },
+                (project, state) =>
+                {
+                    IEnumerable<Module> relatedProjects = SolutionAndProjectUtility.GetProjectPropertiesInternal(project, projectCollection);
+                    foreach (var relatedProject in relatedProjects)
+                    {
+                        allProjects.Add(relatedProject);
+                    }
+                });
+
+            return allProjects;
+        }
+
+        private static string HandleFilteredSolutionFile(string solutionFilterFilePath, out string rootDirectory)
```

Comment:
Maybe 
```suggestion
        private static string HandleFilteredSolutionFilePath(string solutionFilterFilePath, out string rootDirectory)
```
would be more representative?